corehq/apps/sso/models.py

```diff
@@ -1 +1,105 @@
+    This stores the information necessary to make a SAML request to an external
+    IdP. Right now this process supports Azure AD and the plan is to add
+    support for other identity provider types in the future.
+    """
+    # these three fields must only ever be editable by Accounting admins
+    name = models.CharField(max_length=128)
+    slug = models.CharField(max_length=256, db_index=True, unique=True)
+    idp_type = models.CharField(
+        max_length=10,
+        default=IdentityProviderType.AZURE_AD,
+        choices=IdentityProviderType.CHOICES,
+    )
+
+    # whether an IdP is editable by its BillingAccount owner
+    # (it will always be editable by accounting admins)
+    is_editable = models.BooleanField(default=False)
+
+    # whether an IdP is actively in use as an authentication method on HQ
+    is_active = models.BooleanField(default=False)
+
+    # the enterprise admins of this account will be able to edit the SAML
+    # configuration fields
+    owner = models.ForeignKey(BillingAccount, on_delete=models.PROTECT)
+
+    # these are fields required by the external IdP to form a SAML request
+    entity_id = models.TextField(blank=True, null=True)
+    login_url = models.TextField(blank=True, null=True)
+    logout_url = models.TextField(blank=True, null=True)
+    idp_cert_public = models.TextField(blank=True, null=True)
+
+    # the date the IdP's SAML signing certificate expires.
+    # this will be filled out by enterprise admins
+    date_idp_cert_expiration = models.DateTimeField(blank=True, null=True)
+
+    # as the service provider, this will store our x509 certificates and
+    # will be renewed automatically by a periodic task
+    sp_cert_public = models.TextField(blank=True, null=True)
+    sp_cert_private = models.TextField(blank=True, null=True)
+    date_sp_cert_expiration = models.DateTimeField(blank=True, null=True)
+
+    # as the x509 certificate expires, we need to provide the IdP with our next
+    # "rollover" cert to prepare the IdP for the transfer
+    sp_rollover_cert_public = models.TextField(blank=True, null=True)
+    sp_rollover_cert_private = models.TextField(blank=True, null=True)
+    date_sp_rollover_cert_expiration = models.DateTimeField(blank=True, null=True)
+
+    # for auditing purposes
+    created_on = models.DateTimeField(auto_now_add=True)
+    created_by = models.EmailField()
+    last_modified = models.DateTimeField(auto_now=True)
+    last_modified_by = models.EmailField()
+
+    class Meta(object):
+        app_label = 'sso'
+
+    def __str__(self):
+        return f"{self.name} IdP [{self.idp_type}]"
+
+
+class AuthenticatedEmailDomain(models.Model):
+    """
+    This specifies the email domains that are tied to an Identity Provider and
+    a list of users that would be exempt from SSO.
+    """
+    email_domain = models.CharField(max_length=253, db_index=True, unique=True)
+    identity_provider = models.ForeignKey(IdentityProvider, on_delete=models.PROTECT)
```

Comment:
I always default to using `PROTECT` so that whoever is deleting an object knows that there are other things still dependent on that object so the deletion should be very explicit about making sure everything else is properly considered/addressed.

Reply:
Fair, I think I believe PROTECT meant that the child models would be left behind, not that they would prevent deletion of the parent. I think defaulting to PROTECT is fine